### PR TITLE
ailment specs

### PR DESCRIPTION
**Pull request by karolwartak**

### CODE REVIEW

File: src/main/java/com/miquido/validoctor/multirule/ConditionalPropertyRule.java [hunk 1]
```diff
@@ -27,8 +28,21 @@ public boolean test(PatientType obj) {
   }
 
   @Override
-  public Ailment getAilment() {
-    return rule.getAilment();
+  public Ailment apply(PatientType obj) {
+    if (condition.test(obj)) {
+      return rule.apply(propertyGetter.apply(obj));
+    }
+    return null;
```
Comment (amolkaMiquido) [inline src/main/java/com/miquido/validoctor/multirule/ConditionalPropertyRule.java]: Returning null is bad practice - high possibility of NullPointerException. I suggest to return optional or throw checked exception when there is no Ailment

File: src/main/java/com/miquido/validoctor/ailment/CachingAilmentFactory.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package com.miquido.validoctor.ailment;
+
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * Implementation of AilmentFactory that will cache last stated Ailment and return it in subsequent calls
+ * to {@link AilmentFactory#state(Map)} made with the same argument.
+ */
+public class CachingAilmentFactory implements AilmentFactory {
```
Comment (amolkaMiquido) [inline src/main/java/com/miquido/validoctor/ailment/CachingAilmentFactory.java]: This is gold. Good job

File: src/main/java/com/miquido/validoctor/reducerrule/ReducerRule.java [hunk 3]
```diff
@@ -50,8 +51,19 @@ public boolean test(PatientType obj) {
   }
 
   @Override
-  public Ailment getAilment() {
-    return rule.getAilment();
+  public Ailment apply(PatientType obj) {
+    PropertyType sum = getters.stream().map(getter -> getter.apply(obj)).reduce(reducer).orElse(null);
```
Comment (amolkaMiquido) [inline src/main/java/com/miquido/validoctor/reducerrule/ReducerRule.java]: Due to this line, every rule predicate should start with the null check (it's a common approach in javax.validation) and you should add it to documentation.

File: src/main/java/com/miquido/validoctor/multirule/ConditionalPropertyRule.java (hunk 1 above)
Comment (karolwartak) [reply src/main/java/com/miquido/validoctor/multirule/ConditionalPropertyRule.java]: i do not see why this is an exceptional state? apply's method contract states it should return null if there is no aliment stated. It is called by Validoctor class only so there is no chance of NullPointerException

**Pull request merged**